File: scala-package/core/src/main/scala/ml/dmlc/mxnet/KVStore.scala
package ml.dmlc.mxnet

import ml.dmlc.mxnet.Base._
import org.slf4j.{LoggerFactory, Logger}

/**
 * Key value store interface of MXNet for parameter synchronization.
 * @author Yizhi Liu
 */
object KVStore {

  // group id of scheduler/server/worker
  val GROUP_NODE_SCHEDULER = 1
  val GROUP_NODE_SERVER = 2
  val GROUP_NODE_WORKER = 4

  /**
   * Create a new KVStore. <br />
   * <b>
   * WARNING: it is your responsibility to clear this object through dispose().
   * NEVER rely on the GC strategy
   * </b>
   *
   * @param name : {'local', 'dist'}
   *     The type of KVStore
   *     - local works for multiple devices on a single machine (single process)
   *     - dist works for multi-machines (multiple processes)
   * @return The created KVStore
   */
  def create(name: String = "local"): KVStore = {
    val handle = new KVStoreHandleRef
    checkCall(_LIB.mxKVStoreCreate(name, handle))
    new KVStore(handle.value)
  }
}

// scalastyle:off finalize
class KVStore(private[mxnet] val handle: KVStoreHandle) {
  private val logger: Logger = LoggerFactory.getLogger(classOf[KVStore])
  private var updaterFunc: MXKVStoreUpdater = null
  private var disposed = false

  override protected def finalize(): Unit = {
    dispose()
  }

  /**
   * Release the native memory.
   * The object shall never be used after it is disposed.
   */
  def dispose(): Unit = {
    if (!disposed) {
      _LIB.mxKVStoreFree(handle)
      disposed = true
    }
  }

  /**
   * Initialize a single or a sequence of key-value pairs into the store.
   * For each key, one must init it before push and pull.
   * Only worker 0's (rank == 0) data are used.
   * This function returns after data have been initialized successfully
   *
   * @param keys The keys.
   * @param values The values.
   */
  def init(keys: Array[Int], values: Array[NDArray]): Unit = {
    require(keys.length == values.length, "len(keys) != len(values)")
    val valuePtrs = values.map(_.handle)
    checkCall(_LIB.mxKVStoreInit(handle, keys.length, keys, valuePtrs))
  }

  def init(key: Int, value: NDArray): Unit = {
    init(Array(key), Array(value))
  }

  /**
   * Push a single or a sequence of key-value pairs into the store.
   * Data consistency:
   * 1. this function returns after adding an operator to the engine.
   * 2. push is always called after all previous push and pull on the same key are finished
   * 3. there is no synchronization between workers. One can use _barrier() to sync all workers
   *
   * @param keys Keys
   * @param values  According values
   * @param priority
   *         The priority of the push operation.
   *         The higher the priority, the faster this action is likely
   *         to be executed before other push actions.
   */
  def push(keys: Array[Int], values: Array[NDArray], priority: Int): Unit = {
    require(keys.length == values.length, "len(keys) != len(values)")
    val valuePtrs = values.map(_.handle)
    checkCall(_LIB.mxKVStorePush(handle, keys.length, keys, valuePtrs, priority))
  }

  def push(keys: Array[Int], values: Array[NDArray]): Unit = push(keys, values, 0)

  def push(key: Int, value: NDArray, priority: Int = 0): Unit = {
    push(Array(key), Array(value), priority)
  }

  def push(key: Int, values: Array[NDArray], priority: Int): Unit = {
    val keys = Array.fill(values.length)(key)
    push(keys, values, priority)
  }

  def push(key: Int, values: Array[NDArray]): Unit = {
    push(key, values, 0)
  }

  /**
   * Pull a single value or a sequence of values from the store.
   *
   * Data consistency:
   * 1. this function returns after adding an operator to the engine. But any
   *    further read on out will be blocked until it is finished.
   * 2. pull is always called after all previous push and pull on the same key are finished
   * 3. It pulls the newest value from the store.
   * @param keys Keys
   * @param outs According values
   * @param priority
   *     The priority of the push operation.
   *     The higher the priority, the faster this action is likely
   *     to be executed before other push actions.
   */
  def pull(keys: Array[Int], outs: Array[NDArray], priority: Int): Unit = {
    require(keys.length == outs.length, "len(keys) != len(outs)")
    val outPtrs = outs.map(_.handle)
    checkCall(_LIB.mxKVStorePull(handle, keys.length, keys, outPtrs, priority))
  }

  def pull(keys: Array[Int], outs: Array[NDArray]): Unit = pull(keys, outs, 0)

  def pull(key: Int, out: NDArray, priority: Int = 0): Unit = {
    pull(Array(key), Array(out), priority)
  }

  def pull(key: Int, outs: Array[NDArray], priority: Int): Unit = {
    val keys = Array.fill(outs.length)(key)
    pull(keys, outs, priority)
  }

  def pull(key: Int, outs: Array[NDArray]): Unit = {
    pull(key, outs, 0)
  }

  // Get the type of this kvstore
  def `type`: String = {
    val kvType = new RefString
    checkCall(_LIB.mxKVStoreGetType(handle, kvType))
    kvType.value
  }

  /**
   * Get the number of worker nodes
   * @return The number of worker nodes
   */
  def numWorkers: Int = {
    val size = new RefInt
    checkCall(_LIB.mxKVStoreGetGroupSize(handle, size))
    size.value
  }

  /**
   * Get the rank of this worker node
   * @return The rank of this node, which is in [0, get_num_workers())
   */
  def rank: Int = {
    val rank = new RefInt
    checkCall(_LIB.mxKVStoreGetRank(handle, rank))
    rank.value
  }

  /**
   * Register an optimizer to the store
   * If there are multiple machines, this process (should be a worker node)
   * will pack this optimizer and send it to all servers. It returns after
   * this action is done.
   *
   * @param optimizer the optimizer
   */
  def setOptimizer(optimizer: Optimizer): Unit = {
    val isWorker = new RefInt
    checkCall(_LIB.mxKVStoreIsWorkerNode(isWorker))
    if (`type`.contains("dist") && isWorker.value != 0) {
      val optSerialized = Serializer.getSerializer.serialize(optimizer)
      val cmd = Serializer.encodeBase64String(optSerialized)
      logger.debug("Send optimizer to server: {}", cmd)
      sendCommandToServers(0, cmd)
    } else {
      setUpdater(Optimizer.getUpdater(optimizer))
    }
  }

  /**
   * Set a push updater into the store.
   *
   * This function only changes the local store. Use setOptimizer for
   * multi-machines.
   *
   * @param updater  the updater function
   */
  def setUpdater(updater: MXKVStoreUpdater): Unit = {
    this.updaterFunc = updater
    checkCall(_LIB.mxKVStoreSetUpdater(handle, updaterFunc))
  }

  /**
   * Global barrier among all worker nodes
   *
   * For example, assume there are n machines, we want to let machine 0 first
   * init the values, and then pull the inited value to all machines. Before
   * pulling, we can place a barrier to guarantee that the initialization is
   * finished.
   */
  def barrier(): Unit = {
    checkCall(_LIB.mxKVStoreBarrier(handle))
  }

  def numDeadNode(nodeId: Int): Int = {
    val number = new RefInt
    checkCall(_LIB.mxKVStoreGetNumDeadNode(handle, nodeId, number))
    number.value
  }

  /**
   * Whether to do barrier when the kvstore finalizes
   * @param barrierBeforeExit
   */
  def setBarrierBeforeExit(barrierBeforeExit: Boolean): Unit = {
    val flag: Int = if (barrierBeforeExit) 1 else 0
    checkCall(_LIB.mxKVStoreSetBarrierBeforeExit(handle, flag))
  }

  /**
   * Send a command to all server nodes
   *
   * Send a command to all server nodes, which will make each server node run
   * KVStoreServer.controller
   *
   * This function returns after the command has been executed in all server nodes
   *
   * @param head the head of the command
   * @param body the body of the command
   */
  private def sendCommandToServers(head: Int, body: String): Unit = {
    checkCall(_LIB.mxKVStoreSendCommmandToServers(handle, head, body))
  }
}
// scalastyle:off finalize

File: contrib/clojure-package/src/org/apache/clojure_mxnet/kvstore.clj
;;
;; Licensed to the Apache Software Foundation (ASF) under one or more
;; contributor license agreements.  See the NOTICE file distributed with
;; this work for additional information regarding copyright ownership.
;; The ASF licenses this file to You under the Apache License, Version 2.0
;; (the "License"); you may not use this file except in compliance with
;; the License.  You may obtain a copy of the License at
;;
;;    http://www.apache.org/licenses/LICENSE-2.0
;;
;; Unless required by applicable law or agreed to in writing, software
;; distributed under the License is distributed on an "AS IS" BASIS,
;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
;; See the License for the specific language governing permissions and
;; limitations under the License.
;;

(ns org.apache.clojure-mxnet.kvstore
  (:refer-clojure :exclude [type])
  (:require [clojure.spec.alpha :as spec]
            [org.apache.clojure-mxnet.util :as util]
            [clojure.spec.alpha :as s])
  (:import (org.apache.mxnet KVStore NDArray)))

(defn create
  " Create a new KVStore
   WARNING: it is your responsibility to clear this object through dispose.
   - name : #{local, dist} (default is local)
        The type of KVStore
        - local works for multiple devices on a single machine (single process)
        - dist works for multi-machines (multiple processes)"
  ([name]
   (KVStore/create name))
  ([]
   (create "local")))

(defn dispose
  "Release the native memory.
   The object shall never be used after it is disposed."
  [kvstore]
  (.dispose kvstore))

(s/def ::ks (s/or :string string?
                  :vec-of-string (s/coll-of string? :kind vector?)))
(s/def ::ndarray #(instance? NDArray %))
(s/def ::vs (s/or :ndarray ::ndarray
                  :vec-of-ndarray (s/coll-of ::ndarray :kind vector?)))

(defn init
  "Initialize a single or a sequence of key-value pairs into the store.
    For each key, one must init it before push and pull.
    Only worker 0's (rank == 0) data are used.
    This function returns after data have been initialized successfully
    kvstore - KVstore
    ks - keys (vec or strings or single string)
    vs - values (vec or NDArrays or single ndarry)"
  [kvstore ks vs]
  (util/validate! ::ks ks "Invalid keys")
  (util/validate! ::vs vs "Invalid values")
  (doto kvstore
    (.init (into-array (if (vector? ks) ks [ks]))
           (into-array (if (vector? vs) vs [vs])))))

(s/def ::priority int?)

(defn push
  " Push a single or a sequence of key-value pairs into the store.
    Data consistency:
    1. this function returns after adding an operator to the engine.
    2. push is always called after all previous push and pull on the same key are finished
    3. there is no synchronization between workers. One can use _barrier() to sync all workers

   -ks Keys
   -vs values  According values
   - priority
           The priority of the push operation.
           The higher the priority, the faster this action is likely
           to be executed before other push actions."
  ([kvstore ks vs priority]
   (util/validate! ::ks ks "Invalid keys")
   (util/validate! ::vs vs "Invalid values")
   (util/validate! ::priority priority "Invalid priority")
   (let [store-vals (if (vector? vs) vs [vs])
         store-keys (if (vector? ks) ks (into [] (repeat (count store-vals) ks)))]
     (doto kvstore
       (.push (into-array store-keys)
              (into-array store-vals)
              (int priority)))))
  ([kvstore ks vs]
   (push kvstore ks vs 0)))

(s/def ::outs (s/or :ndarray ::ndarray
                    :vec-of-ndarray (s/coll-of ::ndarray :kind vector?)))

(defn pull
  " Pull a single value or a sequence of values from the store.
    Data consistency:
    1. this function returns after adding an operator to the engine. But any
       further read on out will be blocked until it is finished.
    2. pull is always called after all previous push and pull on the same key are finished
    3. It pulls the newest value from the store.
   - kvstore
   - ks single or vector of (strings)
   - outs single or vector of outs (NDArrays)
   - priority
       The priority of the push operation.
       The higher the priority, the faster this action is likely
       to be executed before other push actions."
  ([kvstore ks outs priority]
   (util/validate! ::ks ks "Invalid keys")
   (util/validate! ::outs outs "Invalid outs")
   (util/validate! ::priority priority "Invalid priority")
   (let [store-vals (if (vector? outs) outs [outs])
         store-keys (if (vector? ks) ks (into [] (repeat (count store-vals) ks)))]
     (doto kvstore
       (.pull (into-array store-keys)
              (into-array store-vals)
              (int priority)))))
  ([kvstore ks outs]
   (pull kvstore ks outs 0)))

(defn type
  "Get the type of the kvstore"
  [kvstore]
  (.type kvstore))

(defn num-workers
  "Get the number of worker nodes"
  [kvstore]
  (.numWorkers kvstore))

(defn rank
  "Get the rank of this worker node
   returns The rank of this node, which is in [0, get_num_workers()) "
  [kvstore]
  (.rank kvstore))

(defn set-optimizer
  "Register an optimizer to the store
   If there are multiple machines, this process (should be a worker node)
   will pack this optimizer and send it to all servers. It returns after
   this action is done"
  [kvstore optimizer]
  (doto kvstore
    (.setOptimizer optimizer)))

(defn barrier
  "Global barrier among all worker nodes
    For example, assume there are n machines, we want to let machine 0 first
    init the values, and then pull the inited value to all machines. Before
    pulling, we can place a barrier to guarantee that the initialization is
    finished."
  [kvstore]
  (doto kvstore
    (.barrier kvstore)))

(defn num-dead-node [kvstore node-id]
  (.numDeadNode kvstore (int node-id)))

(defn set-barrier-before-exit
  " Whether to do barrier when the kvstore finalizes
   - kvstore
   - barrier-before-exit boolean"
  [kvstore barrier-before-exit]
  (doto kvstore
    (.setBarrierBeforeExit barrier-before-exit)))

(s/def ::head int?)
(s/def ::body string?)

(defn send-command-to-servers
  "Send a command to all server nodes
    Send a command to all server nodes, which will make each server node run
    KVStoreServer.controller
    This function returns after the command has been executed in all server nodes
   -kvstore
   -head the head of the command
   - body the body of the command"
  [kvstore head body]
  (util/validate! ::head head "Invalid head")
  (util/validate! ::body body "Invalid body")
  (doto kvstore
    (.sendCommandToServers (int head) body)))

(s/def ::fname string?)

(defn save-optimizer-states
  "Save optimizer (updater) state to file
   - kvstore
   - fname Path to output states file."
  [kvstore fname]
  (util/validate! ::fname fname "Invalid filename")
  (doto kvstore
    (.saveOptimizerStates fname)))

(defn load-optimizer-states
  "Load optimizer (updater) state from file
   - kvstore
   -fname Path to input states file."
  [kvstore fname]
  (util/validate! ::fname fname "Invalid filename")
  (doto kvstore
    (.loadOptimizerStates fname)))
